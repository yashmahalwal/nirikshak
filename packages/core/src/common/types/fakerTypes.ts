import _ from "lodash";
import faker from "faker";
import { FakerString } from "./fakerStrings";

// Type guard: Faker string
export function isFakerString(input: any): input is FakerString {
    if (typeof input !== "string") return false;

    const i = input.indexOf("faker:");
    if (i != 0) return false;

    return _.get(faker, input.slice(6)) instanceof Function;
}

// Faker object type
export type FakerObject = {
    function: FakerString;
    args?: any[];
};

// type guard: Faker object
export function isFakerObject(input: any): input is FakerObject {
    if (!input || typeof input !== "object") return false;

    if ("function" in input && isFakerString(input["function"])) {
        if ("args" in input) {
            return Array.isArray(input["args"]);
        }
        return true;
    }

    return false;
}

// The faker literal can be either of both
export type FakerType = FakerString | FakerObject;

// Type guard: faker type
export function isFakerType(input: any): input is FakerType {
    return isFakerString(input) || isFakerObject(input);
}

// A standard format to rep faker types
// Same as faker object, but optional fields are made compulsory
export type NormalizedFaker = {
    [K in keyof FakerObject]-?: FakerObject[K];
};

// Function to normalize faker literal
export function normalizeFaker(input: FakerType): NormalizedFaker {
    if (isFakerObject(input)) return { args: [], ...input };
    return {
        function: input,
        args: [],
    };
}
