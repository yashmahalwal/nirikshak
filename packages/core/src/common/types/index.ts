export { SchemaHelpers } from "./helpers";
export { isCustomFunction, CustomFunction } from "./custom";
export { isFakerType } from "./fakerTypes";
