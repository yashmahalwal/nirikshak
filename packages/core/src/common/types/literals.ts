/* Literals that can be provided to describe the resource */

// Basic literal type
// Ex: 1, false, null, "myString"
export type Literal = string | number | null | boolean;

// type guard: Literal
export function isLiteral(input: any): input is Literal {
    const c = typeof input;
    switch (c) {
        case "number":
        case "string":
        case "boolean":
            return true;
        default:
            return input === null;
    }
}
