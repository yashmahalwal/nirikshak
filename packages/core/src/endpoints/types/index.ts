export {
    MethodType,
    HeaderMap,
    Cases,
    BodyType,
    BodyInstance,
} from "./helpers";
export { isURLString } from "./urlString";
export { isResourceString } from "./resourceString";
export { isDescription, Description } from "./description";
export { HeaderAndQuery, Inputs } from "./input";
export { HeaderAndStatus, Outputs } from "./output";
