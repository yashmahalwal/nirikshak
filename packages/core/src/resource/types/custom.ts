import { Literal } from "./primitive";

// User need to guarantee the validity at runtime
export type Custom = string;
export interface CustomFunction {
  (): Literal;
}
