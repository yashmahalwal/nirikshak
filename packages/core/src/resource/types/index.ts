import {
    isBaseType,
    isWithModifiersBaseType,
    isWithModifiersResource,
    isOneOfEntries,
    BaseType,
    WithModifiers,
    OneOfEntries,
} from "./helper";

// type guard: ResourceBase
export function isResourceBase(input: any): input is ResourceBase {
    if (!input || typeof input != "object" || Array.isArray(input))
        return false;
    return Object.keys(input).every(
        (key) =>
            isBaseType(input[key]) ||
            isWithModifiersBaseType(input[key]) ||
            isWithModifiersResource(input[key]) ||
            isOneOfEntries(input[key]) ||
            isResourceBase(input[key])
    );
}

// A basic key value pair
export interface ResourceBase {
    [key: string]:
        | BaseType
        | WithModifiers<BaseType>
        | WithModifiers<ResourceBase>
        | OneOfEntries
        // Check for resource base at last as it matches any key
        | ResourceBase;
}

// A resource schema is a resource base along with an identifier field
export interface Resource extends ResourceBase {
    id: Exclude<BaseType, Array<any> | null | boolean>;
}

// type guard: isResource
export function isResource(input: any): input is Resource {
    if (!input || typeof input !== "object") return false;

    return (
        !Array.isArray(input["id"]) &&
        isBaseType(input.id) &&
        isResourceBase(input)
    );
}
