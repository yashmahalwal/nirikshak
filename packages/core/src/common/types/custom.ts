// Structuring similar to faker types
// Custom function string is path to a helper function kept in the working directory

import { Literal } from "./literals";

// Custom function string, starts with custom:
// Ex: custom:random.name
export type CustomFunctionString = string;

// Object format of the custom function
// Ex: {function: "custom:random.name"}
export type CustomFunctionObject = {
    function: CustomFunctionString;
    args?: any[];
};

// A custom function that user provides
// User needs to guarantee the validity at runtime
export interface CustomFunction {
    // Primitive is simply a literal or an array of nested arrays of literals
    // Ex: 1, [1,7], [false, [true, [7,8,false, [null]]]]
    (...args: any[]): Promise<Literal>;
}
// Custom type
export type CustomFunctionType = CustomFunctionString | CustomFunctionObject;

// type guard: CustomFunctionString
export function isCustomFunctionString(
    input: any
): input is CustomFunctionString {
    if (typeof input !== "string") return false;

    return input.startsWith("custom:");
}

// type guard: CustomFunctinObject
export function isCustomFunctionObject(
    input: any
): input is CustomFunctionObject {
    if (!input || typeof input !== "object") return false;

    return (
        "function" in input &&
        isCustomFunctionString(input["function"]) &&
        ("args" in input ? Array.isArray(input["args"]) : true)
    );
}

// type guard: CustomFunction
export function isCustomFunction(input: any): input is CustomFunctionType {
    return isCustomFunctionString(input) || isCustomFunctionObject(input);
}

// Normalized form of custom functions
// Same as CustomFunctionObject but optional fields made compulsort
// Ex: {function : "custom:random.name", args: []}
export type NormalizedCustomFunction = {
    [K in keyof CustomFunctionObject]-?: CustomFunctionObject[K];
};

// Function to normalize custom functions
export function normalizeCustomFunction(
    input: CustomFunctionType
): NormalizedCustomFunction {
    if (isCustomFunctionObject(input)) return { args: [], ...input };
    return {
        function: input,
        args: [],
    };
}
