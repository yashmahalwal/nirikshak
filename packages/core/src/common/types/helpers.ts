import { CustomFunction } from "./custom";
import { Literal, isLiteral } from "./literals";

export interface SchemaHelpers {
    [key: string]: CustomFunction;
}

// Ultimately, a resource fields ends in Primitive type
// Ex: {age: 23}
// Ex: {address: [{zipCode: "370445"}, {zipCode: "121342"}], classes: [1,2,3,4,5]}
export type Primitives = Literal | Array<Primitives>;
export function isPrimitives(input: any): input is Primitives {
    if (Array.isArray(input)) return input.every((i) => isPrimitives(i));
    return isLiteral(input);
}
