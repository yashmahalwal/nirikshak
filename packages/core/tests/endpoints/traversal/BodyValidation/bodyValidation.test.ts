import {
    SchemaHelpers,
    TraversalHelpers,
    BodyType,
    ResourceInstance,
    bodyValidation,
    NodeEntry,
    InputInstance,
    ParsedNode,
} from "../../../../src";
import { CustomFunctionType } from "../../../../src/common/types/custom";

const schemaHelpers: SchemaHelpers = {
    word: async (): Promise<string> => "word word",
    invalid: async () => ({ age: 12 } as any),
};

const traversalHelpers: TraversalHelpers = {
    validate: async (input: any) => input === "secretKeyHash",
    invalid: async () => 12 as any,
};

const resource: ResourceInstance = {
    id: "identification",
    age: 3,
    ages: {
        classes: {
            places: [{ a: 12 }, { b: 13 }],
        },
    },
    purpose: [1, 2, 4],
    lorem: [{ ipsum: 4 }, { ipsum: 5 }],
};

const Bodies: (
    | { type: "body type"; value: BodyType }
    | { type: "custom function"; value: CustomFunctionType }
)[] = [
    {
        type: "body type",
        value: {
            id: "resource:id",
            purpose: "resource:purpose",
        },
    },
    {
        type: "custom function",
        value: "custom:validate",
    },
    {
        type: "body type",
        value: {
            id: "resource:id",
            age: "faker:random.number",
        },
    },
];

const entries: { input: any; output: boolean }[] = [
    {
        input: {
            id: resource.id,
            purpose: resource.purpose,
        },
        output: true,
    },
    { input: "secretKeyHash", output: true },
    {
        input: {
            id: resource.id,
            age: resource.age,
        },
        output: true,
    },
    {
        input: {
            id: resource.id,
        },
        output: false,
    },
    { input: "secretKey", output: false },
];

const InvalidBodies: {
    input: any;
    body: { type: "body type" | "custom function"; value: any };
}[] = [
    {
        input: {
            age: 53,
        },
        body: {
            value: {
                age: "resource:non.existent",
            },
            type: "body type",
        },
    },
    {
        input: {
            ages: [1],
        },
        body: {
            type: "body type",
            value: { ages: ["resource:ages"] },
        },
    },
    {
        input: {
            classes: "purple",
        },
        body: {
            type: "body type",
            value: { classes: "custom:color" },
        },
    },
    {
        input: { purpose: false },
        body: {
            type: "body type",
            value: { purpose: "custom:invalid" },
        },
    },
    {
        input: { a: "1" },
        body: { type: "custom function", value: "custom:color" },
    },
    {
        input: { a: "1" },
        body: { type: "custom function", value: "custom:invalid" },
    },
];

const inputInstance: InputInstance = {
    semantics: {
        query: {
            sort: true,
        },
    },
};

const parsedNode: ParsedNode = {
    caseValue: "POSITIVE",
    method: "GET",
    methodIndex: 0,
    url: "loremIpsumDolorSitAmet",
};

const nodeEntry: NodeEntry = {
    url: parsedNode.url,
    input: {
        semantics: {
            query: {
                sort: true,
            },
        },
    },
    output: [
        {
            semantics: { status: 200 },
            body: "custom:function",
        },
    ],
};

describe(`Validate body`, () => {
    test(`Empty array`, () => {
        expect(
            bodyValidation(
                {},
                [],
                resource,
                schemaHelpers,
                traversalHelpers,
                nodeEntry,
                new Map([[resource.id, resource]]),
                inputInstance,
                parsedNode
            )
        ).resolves.toBe(true);
    });

    test.each(entries)(`Entry: %#`, (entry) => {
        expect(
            bodyValidation(
                entry.input,
                Bodies,
                resource,
                schemaHelpers,
                traversalHelpers,
                nodeEntry,
                new Map([[resource.id, resource]]),
                inputInstance,
                parsedNode
            )
        ).resolves.toBe(entry.output);
    });

    test.each(InvalidBodies)(`Invalid entry: %#`, (entry) => {
        expect(
            bodyValidation(
                entry.input,
                [entry.body],
                resource,
                schemaHelpers,
                traversalHelpers,
                nodeEntry,
                new Map([[resource.id, resource]]),
                inputInstance,
                parsedNode
            )
        ).rejects.toMatchSnapshot();
    });
});
