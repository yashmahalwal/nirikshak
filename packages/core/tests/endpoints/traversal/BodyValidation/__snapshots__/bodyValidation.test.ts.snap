// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate body Invalid entry: 0 1`] = `[Error: Invalid resource instance path: non.existent]`;

exports[`Validate body Invalid entry: 1 1`] = `[Error: Resource string resource:ages resolved to non primitive value]`;

exports[`Validate body Invalid entry: 3 1`] = `[Error: function: custom:color, args: ]`;

exports[`Validate body Invalid entry: 4 1`] = `[Error: Non literal type value from custom:invalid]`;

exports[`Validate body Invalid entry: 4 2`] = `[Error: Could not find traversal helper function: custom:color, args: ]`;

exports[`Validate body Invalid entry: 5 1`] = `[Error: function: custom:invalid, args:  did not return a boolean value]`;
